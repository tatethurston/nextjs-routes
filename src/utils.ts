import { existsSync, readdirSync, statSync } from "fs";
import { join, parse } from "path";

const NEXTJS_NON_ROUTABLE = ["/_app", "/_document", "/_error", "/middleware"];
const DYNAMIC_SEGMENT_RE = /\[(.*?)\]/g;

// istanbul ignore next: io mocking not worthwhile
export function getPagesDirectory(): string | undefined {
  const dirs = ["pages", join("src", "pages")];
  for (const dir of dirs) {
    if (existsSync(dir)) {
      return dir;
    }
  }
}

// istanbul ignore next: io mocking not worthwhile
export function findFiles(entry: string): string[] {
  return readdirSync(entry).flatMap((file) => {
    const filepath = join(entry, file);
    if (
      statSync(filepath).isDirectory() &&
      !filepath.includes("node_modules")
    ) {
      return findFiles(filepath);
    }
    return filepath;
  });
}

type QueryType = "dynamic" | "catch-all" | "optional-catch-all";

interface Route {
  pathname: string;
  query: Record<string, QueryType>;
}

function convertWindowsPathToUnix(file: string): string {
  return file.replace(/\\/g, "/");
}

export function nextRoutes(files: string[], pagesDirectory: string): Route[] {
  const filenames = files
    .map((file) => file.replace(pagesDirectory, ""))
    .map((file) => file.replace(parse(file).ext, ""))
    .map(convertWindowsPathToUnix)
    .filter((file) => !NEXTJS_NON_ROUTABLE.includes(file));

  return filenames.map((filename) => {
    const segments = filename.match(DYNAMIC_SEGMENT_RE) ?? [];
    const query = segments.reduce<Route["query"]>((acc, cur) => {
      const param = cur
        .replace(/\[/g, "")
        .replace(/\]/g, "")
        .replace("...", "");
      let queryType: QueryType = "dynamic";
      if (cur.startsWith("[[")) {
        queryType = "optional-catch-all";
      } else if (cur.startsWith("[...")) {
        queryType = "catch-all";
      }
      acc[param] = queryType;
      return acc;
    }, {});

    const pathWithoutIndexSuffix = filename.replace(/index$/, "");
    const pathWithoutTrailingSlash =
      pathWithoutIndexSuffix.endsWith("/") && pathWithoutIndexSuffix.length > 2
        ? pathWithoutIndexSuffix.slice(0, -1)
        : pathWithoutIndexSuffix;

    return {
      pathname: pathWithoutTrailingSlash,
      query,
    };
  });
}

function getQueryInterface(
  query: Route["query"]
): [query: string, requiredKeys: number] {
  let requiredKeys = 0;
  const keys = Object.entries(query)
    .map(([key, value]) => {
      switch (value) {
        case "dynamic": {
          requiredKeys += 1;
          return `${key}: string`;
        }
        case "catch-all": {
          requiredKeys += 1;
          return `${key}: string[]`;
        }
        case "optional-catch-all": {
          return `${key}?: string[] | undefined`;
        }
        // istanbul ignore next
        default: {
          const _exhaust: never = value;
          return _exhaust;
        }
      }
    })
    .join("; ");

  return [`{ ${keys} }`, requiredKeys];
}

export function generate(routes: Route[]): string {
  return `\
// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
// Run \`npx nextjs-routes\` to regenerate this file.
/* eslint-disable */

// prettier-ignore
declare module "nextjs-routes" {
  export type Route =
    ${
      !routes.length
        ? "never"
        : `| ${routes
            .map((route) => {
              const [query, requiredKeys] = getQueryInterface(route.query);
              if (requiredKeys > 0) {
                return `{ pathname: "${route.pathname}"; query: Query<${query}> }`;
              } else {
                return `{ pathname: "${route.pathname}"; query?: Query | undefined }`;
              }
            })
            .join("\n    | ")}`
    };

  type Query<Params = {}> = Params & { [key: string]: string | undefined };
}

// prettier-ignore
declare module "next/link" {
  import type { Route } from "nextjs-routes";
  import type { LinkProps as NextLinkProps } from "next/dist/client/link";
  import type { PropsWithChildren, MouseEventHandler } from "react";
  export * from "next/dist/client/link";

  type RouteOrQuery = Route | { query?: { [key: string]: string | undefined } };

  export interface LinkProps extends Omit<NextLinkProps, "href"> {
    href: RouteOrQuery;
  }

  declare function Link(
    props: PropsWithChildren<LinkProps>
  ): DetailedReactHTMLElement<
    {
      onMouseEnter?: MouseEventHandler<Element> | undefined;
      onClick: MouseEventHandler;
      href?: string | undefined;
      ref?: any;
    },
    HTMLElement
  >;

  export default Link;
}

// prettier-ignore
declare module "next/router" {
  import type { Route } from "nextjs-routes";
  import type { NextRouter as Router } from "next/dist/client/router";
  export * from "next/dist/client/router";
  export { default } from "next/dist/client/router";

  type TransitionOptions = Parameters<Router["push"]>[2];

  type Pathname = Route["pathname"];

  type QueryForPathname = {
    [K in Route as K["pathname"]]: Exclude<K["query"], undefined>;
  };

  type RouteOrQuery = Route | { query: { [key: string]: string | undefined } };

  export interface NextRouter<P extends Pathname = Pathname>
    extends Omit<Router, "push" | "replace"> {
    pathname: P;
    route: P;
    query: QueryForPathname[P];
    push(
      url: RouteOrQuery,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: RouteOrQuery,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
  }

  export function useRouter<P extends Pathname>(): NextRouter<P>;
}
`;
}
