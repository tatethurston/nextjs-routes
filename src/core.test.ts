import { writeFileSync } from "fs";
import { cli, nextRoutes } from "./core.js";
import { findFiles, getPagesDirectory } from "./utils.js";

// eslint-disable-next-line @typescript-eslint/no-unsafe-return
jest.mock("fs", () => ({
  ...jest.requireActual("fs"),
  writeFileSync: jest.fn(),
}));
const writeFileSyncMock = writeFileSync as jest.Mock;

// eslint-disable-next-line @typescript-eslint/no-unsafe-return
jest.mock("./utils.js", () => ({
  ...jest.requireActual("./utils.js"),
  findFiles: jest.fn(),
  getPagesDirectory: jest.fn(),
}));
const getPagesDirectoryMock = getPagesDirectory as jest.Mock;
const findFilesMock = findFiles as jest.Mock;

describe("nextRoutes", () => {
  it("transforms windows paths", () => {
    const pages = ["src\\pages\\[foo]\\bar\\index.ts"];
    const { pathname } = nextRoutes(pages, "src\\pages")[0];
    expect(pathname).toEqual("/[foo]/bar");
  });

  it("colocated test files", () => {
    const pages = ["pages/index.tsx", "pages/index.test.tsx"];
    expect(nextRoutes(pages, "pages")).toMatchInlineSnapshot(`
      [
        {
          "pathname": "/",
          "query": {},
        },
      ]
    `);
  });
});

describe("route generation", () => {
  it("no routes", () => {
    getPagesDirectoryMock.mockReturnValueOnce("pages");
    findFilesMock.mockReturnValueOnce([]);
    cli();
    expect(writeFileSyncMock.mock.calls).toMatchSnapshot();
  });

  it("typescript", () => {
    getPagesDirectoryMock.mockReturnValueOnce("pages");
    findFilesMock.mockReturnValueOnce([
      "pages/404.ts",
      "pages/[foo].ts",
      "pages/[foo]/[bar]/[baz].ts",
      "pages/[foo]/bar/[baz].ts",
      "pages/[foo]/bar/[baz]/foo/[bar].ts",
      "pages/[foo]/baz.ts",
      "pages/_app.ts",
      "pages/middleware.ts",
      "pages/_debug.ts",
      "pages/_debug/health-check.ts",
      "pages/_document.ts",
      "pages/_error.ts",
      "pages/_error/index.ts",
      "pages/api/[[...segments]].ts",
      "pages/api/[...segments].ts",
      "pages/api/bar.ts",
      "pages/foo/[slug].ts",
      "pages/index.ts",
      "pages/not-found.ts",
      "pages/settings/bars/[bar].ts",
      "pages/settings/bars/[bar]/baz.ts",
      "pages/settings/foo.ts",
      "pages/settings/index.ts",
    ]);
    cli();
    expect(writeFileSyncMock.mock.calls).toMatchSnapshot();
  });

  const consoleError = jest
    .spyOn(console, "error")
    .mockReturnValueOnce(undefined);

  const processExit = jest
    .spyOn(process, "exit")
    .mockReturnValueOnce(undefined as never);

  it("missing pages directory", () => {
    getPagesDirectoryMock.mockReturnValueOnce(undefined);
    cli();
    expect(consoleError.mock.calls).toMatchInlineSnapshot(`
      [
        [
          "[nextjs-routes] Could not find a Next.js pages directory. Expected to find either pages(1) or src/pages(2).

        1. https://nextjs.org/docs/basic-features/pages
        2. https://nextjs.org/docs/advanced-features/src-directory
        ",
        ],
      ]
    `);
    expect(processExit).toHaveBeenCalledWith(1);
  });
});
