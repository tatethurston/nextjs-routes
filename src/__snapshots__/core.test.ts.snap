// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`route generation configuration i18n 1`] = `
[
  [
    "nextjs-routes.d.ts",
    "// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
// This file will be automatically regenerated when your Next.js server is running.
/* eslint-disable */

// prettier-ignore
declare module "nextjs-routes" {
  export type Route =
    | { pathname: "/"; query?: Query | undefined };

  type Query<Params = {}> = Params & {
    [key: string]: string | string[] | undefined;
  };

  type QueryForPathname = {
    [K in Route as K["pathname"]]: Exclude<K["query"], undefined>;
  };

  export type RoutedQuery<P extends Route["pathname"]> = QueryForPathname[P];

  export type Locale = 
      | "en-US"
    | "fr"
    | "nl-NL";

  /**
   * A typesafe utility function for generating paths in your application.
   *
   * route({ pathname: "/foos/[foo]", query: { foo: "bar" }}) will produce "/foos/bar".
   */
  export declare function route(r: Route): string;
}

// prettier-ignore
declare module "next/link" {
  import type { Route } from "nextjs-routes";
  import type { LinkProps as NextLinkProps } from "next/dist/client/link";
  import type { PropsWithChildren, MouseEventHandler } from "react";
  export * from "next/dist/client/link";

  type Query = { query?: { [key: string]: string | string[] | undefined } };
  type StaticRoute = Exclude<Route, { query: any }>["pathname"];

  export interface LinkProps<Href extends Route | Query = Route | Query>
    extends Omit<NextLinkProps, "href" | "locale"> {
    href: Href;
    locale?: Locale | false;
  }

  type LinkReactElement = DetailedReactHTMLElement<
    {
      onMouseEnter?: MouseEventHandler<Element> | undefined;
      onClick: MouseEventHandler;
      href?: string | undefined;
      ref?: any;
    },
    HTMLElement
  >;

  declare function Link(
    props: PropsWithChildren<LinkProps<Route>>
  ): LinkReactElement;
  declare function Link(
    props: PropsWithChildren<LinkProps<StaticRoute>>
  ): LinkReactElement;
  declare function Link(
    props: PropsWithChildren<LinkProps<Query>>
  ): LinkReactElement;

  export default Link;
}

// prettier-ignore
declare module "next/router" {
  import type { Locale, Route, RoutedQuery } from "nextjs-routes";
  import type { NextRouter as Router } from "next/dist/client/router";
  export * from "next/dist/client/router";
  export { default } from "next/dist/client/router";

  type NextTransitionOptions = NonNullable<Parameters<Router["push"]>[2]>;
  type StaticRoute = Exclude<Route, { query: any }>["pathname"];

  interface TransitionOptions extends Omit<NextTransitionOptions, "locale"> {
    locale?: Locale | false;
  };

  export interface NextRouter<P extends Route["pathname"] = Route["pathname"]>
    extends Omit<
      Router,
      | "push"
      | "replace"
      | "locale"
      | "locales"
      | "defaultLocale"
      | "domainLocales"
    > {
    defaultLocale: "en-US";
    domainLocales: [
      {
        "domain": "example.nl",
        "defaultLocale": "nl-NL",
        "locales": [
          "nl-BE"
        ]
      },
      {
        "domain": "example.fr",
        "defaultLocale": "fr",
        "http": true
      }
    ];
    locale: Locale;
    locales: [
      "en-US",
      "fr",
      "nl-NL"
    ];
    pathname: P;
    push(
      url: Route,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    push(
      url: StaticRoute,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    push(
      url: { query: { [key: string]: string | string[] | undefined } },
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    query: RoutedQuery<P>;
    replace(
      url: Route,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: StaticRoute,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: { query: { [key: string]: string | string[] | undefined } },
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    route: P;
  }

  export function useRouter<P extends Route["pathname"]>(): NextRouter<P>;
}
",
  ],
]
`;

exports[`route generation configuration outDir 1`] = `
[
  [
    "src/nextjs-routes.d.ts",
    "// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
// This file will be automatically regenerated when your Next.js server is running.
/* eslint-disable */

// prettier-ignore
declare module "nextjs-routes" {
  export type Route =
    | { pathname: "/404"; query?: Query | undefined };

  type Query<Params = {}> = Params & {
    [key: string]: string | string[] | undefined;
  };

  type QueryForPathname = {
    [K in Route as K["pathname"]]: Exclude<K["query"], undefined>;
  };

  export type RoutedQuery<P extends Route["pathname"]> = QueryForPathname[P];

  export type Locale = undefined;

  /**
   * A typesafe utility function for generating paths in your application.
   *
   * route({ pathname: "/foos/[foo]", query: { foo: "bar" }}) will produce "/foos/bar".
   */
  export declare function route(r: Route): string;
}

// prettier-ignore
declare module "next/link" {
  import type { Route } from "nextjs-routes";
  import type { LinkProps as NextLinkProps } from "next/dist/client/link";
  import type { PropsWithChildren, MouseEventHandler } from "react";
  export * from "next/dist/client/link";

  type Query = { query?: { [key: string]: string | string[] | undefined } };
  type StaticRoute = Exclude<Route, { query: any }>["pathname"];

  export interface LinkProps<Href extends Route | Query = Route | Query>
    extends Omit<NextLinkProps, "href" | "locale"> {
    href: Href;
    locale?: false;
  }

  type LinkReactElement = DetailedReactHTMLElement<
    {
      onMouseEnter?: MouseEventHandler<Element> | undefined;
      onClick: MouseEventHandler;
      href?: string | undefined;
      ref?: any;
    },
    HTMLElement
  >;

  declare function Link(
    props: PropsWithChildren<LinkProps<Route>>
  ): LinkReactElement;
  declare function Link(
    props: PropsWithChildren<LinkProps<StaticRoute>>
  ): LinkReactElement;
  declare function Link(
    props: PropsWithChildren<LinkProps<Query>>
  ): LinkReactElement;

  export default Link;
}

// prettier-ignore
declare module "next/router" {
  import type { Locale, Route, RoutedQuery } from "nextjs-routes";
  import type { NextRouter as Router } from "next/dist/client/router";
  export * from "next/dist/client/router";
  export { default } from "next/dist/client/router";

  type NextTransitionOptions = NonNullable<Parameters<Router["push"]>[2]>;
  type StaticRoute = Exclude<Route, { query: any }>["pathname"];

  interface TransitionOptions extends Omit<NextTransitionOptions, "locale"> {
    locale?: false;
  };

  export interface NextRouter<P extends Route["pathname"] = Route["pathname"]>
    extends Omit<
      Router,
      | "push"
      | "replace"
      | "locale"
      | "locales"
      | "defaultLocale"
      | "domainLocales"
    > {
    defaultLocale?: undefined;
    domainLocales?: undefined;
    locale?: Locale;
    locales?: undefined;
    pathname: P;
    push(
      url: Route,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    push(
      url: StaticRoute,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    push(
      url: { query: { [key: string]: string | string[] | undefined } },
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    query: RoutedQuery<P>;
    replace(
      url: Route,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: StaticRoute,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: { query: { [key: string]: string | string[] | undefined } },
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    route: P;
  }

  export function useRouter<P extends Route["pathname"]>(): NextRouter<P>;
}
",
  ],
]
`;

exports[`route generation configuration pageExtensions configured 1`] = `
[
  [
    "nextjs-routes.d.ts",
    "// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
// This file will be automatically regenerated when your Next.js server is running.
/* eslint-disable */

// prettier-ignore
declare module "nextjs-routes" {
  export type Route =
    | { pathname: "/404"; query?: Query | undefined }
    | { pathname: "/"; query?: Query | undefined }
    | { pathname: "/foo"; query?: Query | undefined };

  type Query<Params = {}> = Params & {
    [key: string]: string | string[] | undefined;
  };

  type QueryForPathname = {
    [K in Route as K["pathname"]]: Exclude<K["query"], undefined>;
  };

  export type RoutedQuery<P extends Route["pathname"]> = QueryForPathname[P];

  export type Locale = undefined;

  /**
   * A typesafe utility function for generating paths in your application.
   *
   * route({ pathname: "/foos/[foo]", query: { foo: "bar" }}) will produce "/foos/bar".
   */
  export declare function route(r: Route): string;
}

// prettier-ignore
declare module "next/link" {
  import type { Route } from "nextjs-routes";
  import type { LinkProps as NextLinkProps } from "next/dist/client/link";
  import type { PropsWithChildren, MouseEventHandler } from "react";
  export * from "next/dist/client/link";

  type Query = { query?: { [key: string]: string | string[] | undefined } };
  type StaticRoute = Exclude<Route, { query: any }>["pathname"];

  export interface LinkProps<Href extends Route | Query = Route | Query>
    extends Omit<NextLinkProps, "href" | "locale"> {
    href: Href;
    locale?: false;
  }

  type LinkReactElement = DetailedReactHTMLElement<
    {
      onMouseEnter?: MouseEventHandler<Element> | undefined;
      onClick: MouseEventHandler;
      href?: string | undefined;
      ref?: any;
    },
    HTMLElement
  >;

  declare function Link(
    props: PropsWithChildren<LinkProps<Route>>
  ): LinkReactElement;
  declare function Link(
    props: PropsWithChildren<LinkProps<StaticRoute>>
  ): LinkReactElement;
  declare function Link(
    props: PropsWithChildren<LinkProps<Query>>
  ): LinkReactElement;

  export default Link;
}

// prettier-ignore
declare module "next/router" {
  import type { Locale, Route, RoutedQuery } from "nextjs-routes";
  import type { NextRouter as Router } from "next/dist/client/router";
  export * from "next/dist/client/router";
  export { default } from "next/dist/client/router";

  type NextTransitionOptions = NonNullable<Parameters<Router["push"]>[2]>;
  type StaticRoute = Exclude<Route, { query: any }>["pathname"];

  interface TransitionOptions extends Omit<NextTransitionOptions, "locale"> {
    locale?: false;
  };

  export interface NextRouter<P extends Route["pathname"] = Route["pathname"]>
    extends Omit<
      Router,
      | "push"
      | "replace"
      | "locale"
      | "locales"
      | "defaultLocale"
      | "domainLocales"
    > {
    defaultLocale?: undefined;
    domainLocales?: undefined;
    locale?: Locale;
    locales?: undefined;
    pathname: P;
    push(
      url: Route,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    push(
      url: StaticRoute,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    push(
      url: { query: { [key: string]: string | string[] | undefined } },
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    query: RoutedQuery<P>;
    replace(
      url: Route,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: StaticRoute,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: { query: { [key: string]: string | string[] | undefined } },
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    route: P;
  }

  export function useRouter<P extends Route["pathname"]>(): NextRouter<P>;
}
",
  ],
]
`;

exports[`route generation configuration pageExtensions default 1`] = `
[
  [
    "nextjs-routes.d.ts",
    "// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
// This file will be automatically regenerated when your Next.js server is running.
/* eslint-disable */

// prettier-ignore
declare module "nextjs-routes" {
  export type Route =
    | { pathname: "/404"; query?: Query | undefined };

  type Query<Params = {}> = Params & {
    [key: string]: string | string[] | undefined;
  };

  type QueryForPathname = {
    [K in Route as K["pathname"]]: Exclude<K["query"], undefined>;
  };

  export type RoutedQuery<P extends Route["pathname"]> = QueryForPathname[P];

  export type Locale = undefined;

  /**
   * A typesafe utility function for generating paths in your application.
   *
   * route({ pathname: "/foos/[foo]", query: { foo: "bar" }}) will produce "/foos/bar".
   */
  export declare function route(r: Route): string;
}

// prettier-ignore
declare module "next/link" {
  import type { Route } from "nextjs-routes";
  import type { LinkProps as NextLinkProps } from "next/dist/client/link";
  import type { PropsWithChildren, MouseEventHandler } from "react";
  export * from "next/dist/client/link";

  type Query = { query?: { [key: string]: string | string[] | undefined } };
  type StaticRoute = Exclude<Route, { query: any }>["pathname"];

  export interface LinkProps<Href extends Route | Query = Route | Query>
    extends Omit<NextLinkProps, "href" | "locale"> {
    href: Href;
    locale?: false;
  }

  type LinkReactElement = DetailedReactHTMLElement<
    {
      onMouseEnter?: MouseEventHandler<Element> | undefined;
      onClick: MouseEventHandler;
      href?: string | undefined;
      ref?: any;
    },
    HTMLElement
  >;

  declare function Link(
    props: PropsWithChildren<LinkProps<Route>>
  ): LinkReactElement;
  declare function Link(
    props: PropsWithChildren<LinkProps<StaticRoute>>
  ): LinkReactElement;
  declare function Link(
    props: PropsWithChildren<LinkProps<Query>>
  ): LinkReactElement;

  export default Link;
}

// prettier-ignore
declare module "next/router" {
  import type { Locale, Route, RoutedQuery } from "nextjs-routes";
  import type { NextRouter as Router } from "next/dist/client/router";
  export * from "next/dist/client/router";
  export { default } from "next/dist/client/router";

  type NextTransitionOptions = NonNullable<Parameters<Router["push"]>[2]>;
  type StaticRoute = Exclude<Route, { query: any }>["pathname"];

  interface TransitionOptions extends Omit<NextTransitionOptions, "locale"> {
    locale?: false;
  };

  export interface NextRouter<P extends Route["pathname"] = Route["pathname"]>
    extends Omit<
      Router,
      | "push"
      | "replace"
      | "locale"
      | "locales"
      | "defaultLocale"
      | "domainLocales"
    > {
    defaultLocale?: undefined;
    domainLocales?: undefined;
    locale?: Locale;
    locales?: undefined;
    pathname: P;
    push(
      url: Route,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    push(
      url: StaticRoute,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    push(
      url: { query: { [key: string]: string | string[] | undefined } },
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    query: RoutedQuery<P>;
    replace(
      url: Route,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: StaticRoute,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: { query: { [key: string]: string | string[] | undefined } },
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    route: P;
  }

  export function useRouter<P extends Route["pathname"]>(): NextRouter<P>;
}
",
  ],
]
`;

exports[`route generation no routes 1`] = `
[
  [
    "nextjs-routes.d.ts",
    "// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
// This file will be automatically regenerated when your Next.js server is running.
/* eslint-disable */

// prettier-ignore
declare module "nextjs-routes" {
  export type Route =
    never;

  type Query<Params = {}> = Params & {
    [key: string]: string | string[] | undefined;
  };

  type QueryForPathname = {
    [K in Route as K["pathname"]]: Exclude<K["query"], undefined>;
  };

  export type RoutedQuery<P extends Route["pathname"]> = QueryForPathname[P];

  export type Locale = undefined;

  /**
   * A typesafe utility function for generating paths in your application.
   *
   * route({ pathname: "/foos/[foo]", query: { foo: "bar" }}) will produce "/foos/bar".
   */
  export declare function route(r: Route): string;
}

// prettier-ignore
declare module "next/link" {
  import type { Route } from "nextjs-routes";
  import type { LinkProps as NextLinkProps } from "next/dist/client/link";
  import type { PropsWithChildren, MouseEventHandler } from "react";
  export * from "next/dist/client/link";

  type Query = { query?: { [key: string]: string | string[] | undefined } };
  type StaticRoute = Exclude<Route, { query: any }>["pathname"];

  export interface LinkProps<Href extends Route | Query = Route | Query>
    extends Omit<NextLinkProps, "href" | "locale"> {
    href: Href;
    locale?: false;
  }

  type LinkReactElement = DetailedReactHTMLElement<
    {
      onMouseEnter?: MouseEventHandler<Element> | undefined;
      onClick: MouseEventHandler;
      href?: string | undefined;
      ref?: any;
    },
    HTMLElement
  >;

  declare function Link(
    props: PropsWithChildren<LinkProps<Route>>
  ): LinkReactElement;
  declare function Link(
    props: PropsWithChildren<LinkProps<StaticRoute>>
  ): LinkReactElement;
  declare function Link(
    props: PropsWithChildren<LinkProps<Query>>
  ): LinkReactElement;

  export default Link;
}

// prettier-ignore
declare module "next/router" {
  import type { Locale, Route, RoutedQuery } from "nextjs-routes";
  import type { NextRouter as Router } from "next/dist/client/router";
  export * from "next/dist/client/router";
  export { default } from "next/dist/client/router";

  type NextTransitionOptions = NonNullable<Parameters<Router["push"]>[2]>;
  type StaticRoute = Exclude<Route, { query: any }>["pathname"];

  interface TransitionOptions extends Omit<NextTransitionOptions, "locale"> {
    locale?: false;
  };

  export interface NextRouter<P extends Route["pathname"] = Route["pathname"]>
    extends Omit<
      Router,
      | "push"
      | "replace"
      | "locale"
      | "locales"
      | "defaultLocale"
      | "domainLocales"
    > {
    defaultLocale?: undefined;
    domainLocales?: undefined;
    locale?: Locale;
    locales?: undefined;
    pathname: P;
    push(
      url: Route,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    push(
      url: StaticRoute,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    push(
      url: { query: { [key: string]: string | string[] | undefined } },
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    query: RoutedQuery<P>;
    replace(
      url: Route,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: StaticRoute,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: { query: { [key: string]: string | string[] | undefined } },
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    route: P;
  }

  export function useRouter<P extends Route["pathname"]>(): NextRouter<P>;
}
",
  ],
]
`;

exports[`route generation typescript 1`] = `
[
  [
    "nextjs-routes.d.ts",
    "// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
// This file will be automatically regenerated when your Next.js server is running.
/* eslint-disable */

// prettier-ignore
declare module "nextjs-routes" {
  export type Route =
    | { pathname: "/404"; query?: Query | undefined }
    | { pathname: "/[foo]"; query: Query<{ "foo": string }> }
    | { pathname: "/[foo]/[bar]/[baz]"; query: Query<{ "foo": string; "bar": string; "baz": string }> }
    | { pathname: "/[foo]/bar/[baz]"; query: Query<{ "foo": string; "baz": string }> }
    | { pathname: "/[foo]/bar/[baz]/foo/[bar]"; query: Query<{ "foo": string; "baz": string; "bar": string }> }
    | { pathname: "/[foo]/baz"; query: Query<{ "foo": string }> }
    | { pathname: "/_debug"; query?: Query | undefined }
    | { pathname: "/_debug/health-check"; query?: Query | undefined }
    | { pathname: "/api/[[...segments]]"; query?: Query | undefined }
    | { pathname: "/api/[...segments]"; query: Query<{ "segments": string[] }> }
    | { pathname: "/api/bar"; query?: Query | undefined }
    | { pathname: "/foo/[slug]"; query: Query<{ "slug": string }> }
    | { pathname: "/"; query?: Query | undefined }
    | { pathname: "/not-found"; query?: Query | undefined }
    | { pathname: "/settings/bars/[bar]"; query: Query<{ "bar": string }> }
    | { pathname: "/settings/bars/[bar]/baz"; query: Query<{ "bar": string }> }
    | { pathname: "/settings/foo"; query?: Query | undefined }
    | { pathname: "/settings"; query?: Query | undefined };

  type Query<Params = {}> = Params & {
    [key: string]: string | string[] | undefined;
  };

  type QueryForPathname = {
    [K in Route as K["pathname"]]: Exclude<K["query"], undefined>;
  };

  export type RoutedQuery<P extends Route["pathname"]> = QueryForPathname[P];

  export type Locale = undefined;

  /**
   * A typesafe utility function for generating paths in your application.
   *
   * route({ pathname: "/foos/[foo]", query: { foo: "bar" }}) will produce "/foos/bar".
   */
  export declare function route(r: Route): string;
}

// prettier-ignore
declare module "next/link" {
  import type { Route } from "nextjs-routes";
  import type { LinkProps as NextLinkProps } from "next/dist/client/link";
  import type { PropsWithChildren, MouseEventHandler } from "react";
  export * from "next/dist/client/link";

  type Query = { query?: { [key: string]: string | string[] | undefined } };
  type StaticRoute = Exclude<Route, { query: any }>["pathname"];

  export interface LinkProps<Href extends Route | Query = Route | Query>
    extends Omit<NextLinkProps, "href" | "locale"> {
    href: Href;
    locale?: false;
  }

  type LinkReactElement = DetailedReactHTMLElement<
    {
      onMouseEnter?: MouseEventHandler<Element> | undefined;
      onClick: MouseEventHandler;
      href?: string | undefined;
      ref?: any;
    },
    HTMLElement
  >;

  declare function Link(
    props: PropsWithChildren<LinkProps<Route>>
  ): LinkReactElement;
  declare function Link(
    props: PropsWithChildren<LinkProps<StaticRoute>>
  ): LinkReactElement;
  declare function Link(
    props: PropsWithChildren<LinkProps<Query>>
  ): LinkReactElement;

  export default Link;
}

// prettier-ignore
declare module "next/router" {
  import type { Locale, Route, RoutedQuery } from "nextjs-routes";
  import type { NextRouter as Router } from "next/dist/client/router";
  export * from "next/dist/client/router";
  export { default } from "next/dist/client/router";

  type NextTransitionOptions = NonNullable<Parameters<Router["push"]>[2]>;
  type StaticRoute = Exclude<Route, { query: any }>["pathname"];

  interface TransitionOptions extends Omit<NextTransitionOptions, "locale"> {
    locale?: false;
  };

  export interface NextRouter<P extends Route["pathname"] = Route["pathname"]>
    extends Omit<
      Router,
      | "push"
      | "replace"
      | "locale"
      | "locales"
      | "defaultLocale"
      | "domainLocales"
    > {
    defaultLocale?: undefined;
    domainLocales?: undefined;
    locale?: Locale;
    locales?: undefined;
    pathname: P;
    push(
      url: Route,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    push(
      url: StaticRoute,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    push(
      url: { query: { [key: string]: string | string[] | undefined } },
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    query: RoutedQuery<P>;
    replace(
      url: Route,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: StaticRoute,
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    replace(
      url: { query: { [key: string]: string | string[] | undefined } },
      as?: string,
      options?: TransitionOptions
    ): Promise<boolean>;
    route: P;
  }

  export function useRouter<P extends Route["pathname"]>(): NextRouter<P>;
}
",
  ],
]
`;
